Allow license rules to require the presence of certain defining keywords

mrombout opened the pull request:
Fixes #2637

This PR depends on https://github.com/nexB/scancode-toolkit/pull/2765/commits/97a1a57eec13566328c26e67df6af1bb409144a9, contained in https://github.com/nexB/scancode-toolkit/pull/2765. I have cherry-picked that commit into this PR as 2428723959a2cdc488148e5ecc88807a5d358ebe.

### Tasks

* [x] Reviewed [contribution guidelines](https://github.com/nexB/scancode-toolkit/blob/develop/CONTRIBUTING.rst)
* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] Tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
  Run [tests](https://scancode-toolkit.readthedocs.io/en/latest/contribute/contrib_dev.html#running-tests) locally to check for errors. 
* [x] Commits are in uniquely-named feature branch and has no merge conflicts 📁

mrombout committed:
Add feature to set key phrases in rules

A key phrase can be defined in a rule by surrounding one or more
words (tokens) with the `{{` and `}}` characters. Key phrases are
words that **must** be present in order to successfully match.

The key phrases are parsed for each rule during indexing. Then after
matching using the standard matching algorithm a filter removes all
matches that do not have the key phrases present.

---

At the moment sometimes the `ispan` alone can not be relied upon to determine
if key phrases are present.

At one case stopwords and unknown tokens made it appear as if the `ispan` was
more expansive than it should be. This has been solved by checking if the
`qspan` corresponding with the `ispan` intersect with any unknown tokens or
stopwords.

Another problem that is not solved is with the example boto3 example which
should match with `cc-by-nc-sa-4.0` but it now matching with several other
CC licenses. For some reason `Creative Commons Attribution 4.0 International
License` is considered to be matching in the ispan, and not intersecting with
any unknown tokens or stopwords (citation needed).

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive detection of mpl-1.1

License erlangpl-1.1 was detected as mpl-1.1.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive detection of cddl-1.0

License mit was detected as cddl-1.0.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive detection of imagemagick

License apache-2.0 was detected as imagemagick.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive detection of mit

License apache-2.0 was detected as mit.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

dd-jy committed:
Fix false positive where bsd-3-clause is detected as gpl-2.0-plus_and_bootloader-exception

Signed-off-by: Jiyeong Seok <jiyeong.seok@lge.com>

mrombout committed:
Fix false positive detection of cc-by-4.0

License cc-by-sa-4.0 was detected as cc-by-4.0.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive detection of gpl-2.0

License mit was detected as gpl-2.0.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive of bsd-simplified AND gpl-2.0

Non-license text is detected as bsd-simplified AND gpl-2.0.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive of apache-2.0 AND cc-by-sa-4.0

Non-license text was detected as apache-2.0 AND cc-by-sa-4.0.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive detection of ruby

License mit was detected as ruby.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive detection of apache-2.0

License mpl-2.0 was detected as apache-2.0.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive of apache-2.0 OR epl-2.0

License mit was detected as apache-2.0 OR epl-2.0.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Fix false positive of apache-2.0 OR gpl-2.0-plus

License Bouncy Castyle License (mit) is detected
as apache-2.0 OR gpl-2.0-plus WITH
classpath-exception-2.0.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Add unit tests to key_phrase_tokenizer

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Replace key_phrase_spans default to use factory

Co-authored-by: Philippe Ombredanne <pombredanne@gmail.com>
Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

pombredanne committed:
Do not use defaultdict for Query unknowns

Do not use defaultdict for Query.unknowns_by_pos and
Query.stopwords_by_pos. Otherwise there are pernicious side effects to
add new entries in these dctionaries when querying them after their
creation.

Reported-by: Mike Rombout @mrombout
Signed-off-by: Philippe Ombredanne <pombredanne@nexb.com>

mrombout committed:
Explain key phrase syntax in new license rule guide

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Throw InvalidRule when key phrase is not closed

When a key phrase definition is started in a rule (with `{{`) but is
not closed (with `}}`) the rule is considered invalid and an error is
thrown.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Add key phrases to CHANGELOG

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Add name to AUTHORS.rst

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

pombredanne commented on AUTHORS.rst:
```diff
@@ -43,6 +43,7 @@ The following organizations or individuals have contributed to ScanCode:
 - Maximilian Huber @maxhbr 
 - Michael Herzog @mjherzog
 - Michael Rupprecht @michaelrup
+- Mike Rombout @mrombout
```
Can you add `- Jiyeong Seok @dd-jy`   too there?

pombredanne replied:
@dd-jy I assume you are OK?

pombredanne commented on CHANGELOG.rst:
```diff
@@ -49,6 +49,9 @@ License detection:
   - XXXX new license detection rules have been added, and
   - XXXX existing license rules have been updated.
 
+- Key phrases can now be defined in RULEs by surrounding one or more words with
```
BTW, does this also applies to .LICENSE texts? (not I would want to encourage this in license texts...)

mrombout replied:
Definitely not on purpose, I am not sure on `.LICENSE` texts are used in the system. Though, adding the syntax to a `.LICENSE` text does break some tests, I'll have a look at it.

At the moment the key_phrases are added to the index for every rule https://github.com/softsense/scancode-toolkit/blob/issue-2637-allow-license-rules-to-require-the-presence-of-certain-defining-keywords/src/licensedcode/index.py#L334.

pombredanne replied:
License texts are ingested as if they were rules when indexing: https://github.com/nexB/scancode-toolkit/blob/1f7fafdcc74f1110a7812b4ddf84964b4fe67843/src/licensedcode/models.py#L522 so this likely works there too... I checked that there are no dangling {{ braces }} in any .LICENSE so we are fine too

mrombout committed:
Update filter_key_phrase_spans to also check qspan

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout commented on src/licensedcode/data/rules/false-positive_ansible_2.RULE:
```diff
@@ -5,4 +5,4 @@
   # - Apache
   # - CC-BY
-license: {{ license }}
+license: \{\{ license \}\}
```
In this case the `{{ license }}` is not meant to be a key phrase. As a workaround I've "escaped" the `{{` with a `\`, this as not explicit meaning in the parsing code, it simply trips up the regular expression used to find the tags.

pombredanne replied:
IMHO, do not escape things as it looks weird... just remove the braces entirely.

mrombout replied:
Okay, removed them!

mrombout commented on tests/licensedcode/test_match.py:
```diff
@@ -873,10 +1026,10 @@ class TestCollectLicenseMatchTexts(FileBasedTesting):
-            THIS IS FROM {{THE CODEHAUS}} AND CONTRIBUTORS
-            IN NO EVENT SHALL {{THE CODEHAUS}} OR ITS CONTRIBUTORS BE LIABLE
-            EVEN IF ADVISED OF THE {{POSSIBILITY OF SUCH}} DAMAGE
+            Copyright [[some copyright]]
```
Simply replaced them as a workaround to avoid conflict with the key phrase syntax, do the `{{` and `}}` characters have another special meaning here?

pombredanne replied:
> do the {{ and }} characters have another special meaning here?

In a primitive iteration of license detection there was a notion of "template" where `{{10 some text}}` would mean that up to 10 words could be missed and the match still be valid.
- https://github.com/nexB/scancode-toolkit/blob/v1.0.0/src/licensedcode/models.py#L256
- https://github.com/nexB/scancode-toolkit/blob/v1.0.0/src/textcode/analysis.py#L146
- https://github.com/nexB/scancode-toolkit/blob/v1.0.0/src/textcode/analysis.py#L193

This was byzantine and did not work as expected at all!

mrombout replied:
But this feature is not in anymore, is this test still relevant/correct? Since I changed the characters to `[[` and `]]` (which means nothing), and it didn't break :thinking: Same for the other similar ones.

pombredanne replied:
> But this feature is not in anymore, is this test still relevant/correct? Since I changed the characters to [[ and ]] (which means nothing), and it didn't break thinking Same for the other similar ones.

Yes it could be irrelevant but no harm done either to keep it: this is minor so your changes are fine.

mrombout committed:
Add docstring to BasicRule::key_phrase_spans

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout commented on tests/licensedcode/data/datadriven/lic1/boto-boto3-cw-example-creating-alarm.txt.yml:
```diff
@@ -0,0 +1,5 @@
+license_expressions:
```
Should be only `cc-by-nc-sa-4.0`, missing rule.

pombredanne replied:
correct. Adding a new rules for cc-by-nc-sa-4.0 such as these would cure this

- `is_license_notice: yes`

```
   This work is licensed under a {{Creative Commons Attribution-NonCommercial-ShareAlike 4.0}}
   International License (the "License"). You may not use this file except in compliance with the
   License. A copy of the License is located at http://creativecommons.org/licenses/by-nc-sa/4.0/.

   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
   either express or implied. See the License for the specific language governing permissions and
   limitations under the License.
```

and `is_license_reference: yes` and `relevance: 100`
```
{{Creative Commons Attribution-NonCommercial-ShareAlike 4.0}} International License
```

mrombout replied:
Nice thanks, fixed by adding these rules. I've added two rules where one is the license notice and the other is the license reference. Adding the license notice fixed the test, is the license reference for completeness?

pombredanne replied:
>  is the license reference for completeness?

yes!

pombredanne reviewed (changes requested):
Thanks!
See a few comments in line.

mrombout committed:
Add missing rule for cc-by-nc-sa-4.0

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Remove incompatible key phrase syntax from rules

These rules contained `{{` and `}}` characters, but
were never intended to be considered key phrases.

In order to avoid conflicts I've removed the syntax
from these rules.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout committed:
Add @dd-jy to AUTHORS.rst

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

pombredanne reviewed (approved):
This is overall awesome!
I'd like to check if the impact on indexing is not too bad... but this is not a critical process anymore.

pombredanne commented on src/licensedcode/match.py:
```diff
@@ -1389,6 +1395,55 @@ def filter_already_matched_matches(matches, query):
+    """
+    Return a filtered list of kept LicenseMatch matches and a list of
+    discardable matches by removing all matches that do not contain all key
+    phrases required by the rule.
```
What if there is only one match and this does not contain the keyphrases?
Would you discard it, therefore leaving no matches?

mrombout replied:
If the only match does not contain the key phrase wouldn't that indicate a high chance of being a false positive? If so, for my use-case I would rather have no results, but maybe that's not in-line with Scancode philosophy(e.g. not having false negatives)?

pombredanne replied:
> If the only match does not contain the key phrase wouldn't that indicate a high chance of being a false positive?

yes in the general case, but let's take a practical  example:

1. in query I have  `licensed under the AGPL 3.0` and in rule1 I have `AGPL 3.0`
2. in rule2 I have `licensed under the {{AGPL 3.0 license}}`
3. the query will initially match both rules
4. the match to rule1 may be discarded as contained in the rule2 match
5. the match to rule2 may be discarded as missing a key phrase

I am left with no matches yet `licensed under the AGPL 3.0` was very clearly a license notice.

On the other hand if the match to rule2 is discarded as missing a key phrase early and before the containment filter, then we are at least left with the match to rule1?

mrombout replied:
Okay, that makes sense. But unfortunately I was unable to recreate this scenario in a unit test.

Instead I experimented and moved the key phrase filter up (before the containment filter) and 5 of the datadriven tests I've added broke. Looking into that now. E.g. `tests/licensedcode/test_detection_datadriven1.py::TestLicenseDataDriven1::test_detection_datadriven_lic1_alexa_skills_kit_sdk_for_java_txt` now detects `[apache-2.0, apache-2.0, unknown-license-reference]` as opposed to only `apache-2.0`. 

Also, are you suggesting to skip if there's only one rule like some of the other filters?

```
    if len(matches) < 2:
        return matches
```

pombredanne commented on src/licensedcode/match.py:
```diff
@@ -1485,6 +1540,10 @@ def _log(_matches, _discarded, msg):
         all_discarded.extend(discarded)
         _log(matches, discarded, 'HIGH ENOUGH SCORE')
 
+    matches, discarded = filter_key_phrase_spans(matches)
```
I think this filter should be positioned much earlier in the process otherwise here is the risk:
1. there was a smalller match that was contained
2. it is filtered because of containment in a larger match
3. the larger is discarded because of not having keyphrases
4. there is no match left for this region and we now have a false negative

mrombout replied:
I've moved the key phrase filter up in 2c377b3a6bd69920a7dfb68864d3e7f9bf58901b, to right before the first containment filter. I was unable to reproduce this scenario exactly, but some datadriven tests at least demonstrated the effectiveness (`tests/licensedcode/test_detection_datadriven1.py::TestLicenseDataDriven1::test_detection_datadriven_lic1_alexa_skills_kit_sdk_for_java_txt`)

mrombout replied:
@pombredanne Would the above suffice?

pombredanne reviewed (changes requested):
I added some more comments for your review.

mrombout committed:
Move up key phrase filter to improve results

This will prevent false-positives for the following scenario:

1. There are multiple matches, with a smalller match that id contained
2. It is filtered because of containment in the larger match
3. The larger is discarded because of not having keyphrases
4. There is no match left, we now have a false negative

As a practical example:

1. In query there is `licensed under the AGPL 3.0`
2. In rule1 I have `AGPL 3.0`
3. In rule2 I have `licensed under the {{AGPL 3.0 license}}`
3. The query will initially match both rules
4. The match to rule1 may be discarded as contained in the rule2 match
5. The match to rule2 may be discarded as missing a key phrase

By moving the key phrase filter up before the containment filter the
larger match may be filtered out, giving the smaller matches a chance
to stay.

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

mrombout commented on tests/licensedcode/data/datadriven/lic1/alexa-skills-kit-sdk-for-java.txt.yml:
```diff
@@ -1,2 +1,4 @@
 license_expressions:
   - apache-2.0
+  - apache-2.0
+  - unknown-license-reference
```
I'm suspecting that I've added a key phrase to a rule that prevents it from matching, but now also prevents `unknown-license-reference` for being removed as contained.

See which Apache rule it was and see if key phrase can be less strict, or a new rule must be added.

mrombout replied:
`src/licensedcode/data/rules/apache-2.0_151.RULE` would match probably if only it didn't have the word `licensed`. Instead I've added a new rule `apache-2.0_1026.yml` for now.

mrombout committed:
Fix false positive of apache-2.0

Signed-off-by: Mike Rombout <mike.rombout@elastisys.com>

pombredanne reviewed (approved):
LGTM and thank you ++ for this!
